docs/sphinx/user_guide/feature/resource.rst

```diff
@@ -0,0 +1,268 @@
+            only available when using ``Cuda`` or ``Hip`` resources. 
+          * The ``RAJA::resources`` namespace aliases the ``camp::resources`` namespace.
+
+Each resource has a set of underlying functionality that is synonymous across all resource types.  
+
+ ===================== ===============================================
+ Methods               Brief description
+ ===================== ===============================================
+ get_platform          Returns the underlying camp platform
+                       the resource is associated with.
+ get_event             Return an Event object for the resource from
+                       the last resource call.
+ allocate              Allocate data per the resource's given
+                       backend.
+ deallocate            Deallocate data per the resource's given
+                       backend.
+ memcpy                Perform a memory copy from a src location
+                       to a destination location from the
+                       resource's backend.
+ memset                Set memory value per the resourse's
+                       given backend.
+ wait_for              Enqueue a wait on the resource's stream/thread
+                       for a user passed event to occur.
+ ===================== ===============================================
+  
+Each resource type also defines specific backend information/functionality. For example, each
+Cuda resource contains a ``cudaStream_t`` value with an associated get method. See the 
+individual functionality for each resource in ``raja/tpl/camp/include/resource/``.
+
+
+------------
+Type-Erasure
+------------
+
+Resources can be declared in two formats: An erased resource, and a concrete resource. The 
+underlying runtime functionality is the same for both formats. An erased resource allows a user 
+the ability to change the resource backend at runtime. 
+
+Concrete Cuda resource::
+
+    RAJA::resources::Cuda my_cuda_res;
+
+Erased resource::
+
+    if (use_gpu)
+      RAJA::resources::Resource my_res{RAJA::resources::Cuda()};
+    else
+      RAJA::resources::Resource my_res{RAJA::resources::Host()};
+
+
+Memory allocation on resources::
+
+    int* a1 = my_cuda_res.allocate<int>(ARRAY_SIZE);
+    int* a2 = my_res.allocate<int>(ARRAY_SIZE);
+
+If ``use_gpu`` is ``true``, the underlying type of ``my_res`` is a Cuda resource, therefore ``a1`` 
+and ``a2`` will both be allocated on the GPU. If ``use_gpu`` is ``false``, then only ``a1`` is 
+allocated on the GPU, and ``a2`` is allocated on the host.
+
+
+------
+Forall
+------
+
+A resource is an optional argument to a ``RAJA::forall`` call. It is passed as the first argument 
+when the forall is templated on the execution policy. When the execution policy is passed by value,
+the resource is passed as the second argument. This maintains the order of inputs that can be 
+passed to a ``RAJA::forall`` call::
+
+    RAJA::forall<ExecPol>(my_cuda_res, .... )
+
+or::
```

Comment:
Should we leave this version out? It may confuse folks. As far as I know, users stick with the templated (first) version of forall, which I think is the usage we want to promote.

Reply:
We can take it out, but I know that this version is still available to the user from the API so that's why I thought it would be useful to leave in. Leaving it in depends on how strongly we want to encourage users to use the typed version

Reply:
@trws @MrBurmark @artv3 @rchen20 @davidbeckingsale what do you folks think about this?

Reply:
We don't need to have the second form here, but it will show up in the interface, and would be the version one would have to use with a by-value policy if anyone has those anymore.

Reply:
I don't think anyone is using that form, although it is publicly exposed. I vote for removing it from the docs to avoid user confusion.

Reply:
I have not seen the execution policy used as an object in the wild. Perhaps it's better to leave this out of the documentation to avoid confusion?

Reply:
I think it's an advanced feature, probably doesn't go here, but if we have a discussion of extending forall for example then it would make sense to talk about this kind of thing there.  Not saying move it, this can just come out, but that's more what I was eluding to.

Reply:
Sounds good everyone, I've gone ahead and removed it.